File: historian_v2/src/storage/writer.rs
use anyhow::Result;
use std::collections::HashMap;
use std::path::Path;
use std::sync::Arc;
use tokio::io::AsyncWriteExt;
use tokio::sync::Mutex;
use tokio::time::{sleep, Duration};

use crate::config::{BackpressurePolicy, Config};
use crate::storage::codec::{encode_block_index_entry, encode_segment_record, BlockIndexEntry};
use crate::storage::last_value::LastValueStore;
use crate::storage::layout::{compute_partition, index_path, segment_path, shard_for_tag};
use crate::types::Point;

#[derive(Default, Debug, Clone)]
pub struct WriterStats {
    pub accepted_points: u64,
    pub dropped_points: u64,
    pub decode_errors: u64,
}

#[derive(Debug, Clone)]
struct QueueState {
    points: Vec<Point>,
    bytes: usize,
}

#[derive(Clone)]
pub struct HistorianWriter {
    cfg: Config,
    queues: Arc<Mutex<HashMap<String, QueueState>>>,
    stats: Arc<Mutex<WriterStats>>,
    last_values: LastValueStore,
}

impl HistorianWriter {
    pub fn new(cfg: Config, last_values: LastValueStore) -> Self {
        Self {
            cfg,
            queues: Arc::new(Mutex::new(HashMap::new())),
            stats: Arc::new(Mutex::new(WriterStats::default())),
            last_values,
        }
    }

    pub fn spawn_flush_loop(&self) {
        let me = self.clone();
        tokio::spawn(async move {
            loop {
                let _ = me.flush_all().await;
                sleep(Duration::from_millis(me.cfg.flush.flush_interval_ms)).await;
            }
        });
    }

    pub async fn mark_decode_error(&self) {
        let mut s = self.stats.lock().await;
        s.decode_errors += 1;
    }

    pub async fn ingest_batch(&self, points: Vec<Point>) -> Result<()> {
        let mut accepted = Vec::new();
        {
            let mut queues = self.queues.lock().await;
            let mut s = self.stats.lock().await;
            for p in points {
                let part = compute_partition(p.ts_epoch, &self.cfg);
                let shard = shard_for_tag(p.tag_id, self.cfg.storage.shard_count);
                let key = format!("{}/{}/{}", part.day, part.hour, shard);
                let q = queues.entry(key).or_insert(QueueState {
                    points: Vec::new(),
                    bytes: 0,
                });

                if q.points.len() >= self.cfg.flush.max_queue_points {
                    match self.cfg.flush.backpressure_policy {
                        BackpressurePolicy::DropNew => {
                            s.dropped_points += 1;
                            continue;
                        }
                        BackpressurePolicy::DropOldest => {
                            if let Some(old) = q.points.first() {
                                q.bytes = q.bytes.saturating_sub(encode_segment_record(old).len());
                            }
                            if !q.points.is_empty() {
                                q.points.remove(0);
                            }
                            s.dropped_points += 1;
                        }
                    }
                }
                q.bytes += encode_segment_record(&p).len();
                q.points.push(p.clone());
                accepted.push(p);
                s.accepted_points += 1;
            }
        }
        if !accepted.is_empty() {
            self.last_values.update_batch(&accepted).await?;
        }
        if self.should_flush_fast().await {
            self.flush_all().await?;
        }
        Ok(())
    }

    async fn should_flush_fast(&self) -> bool {
        let queues = self.queues.lock().await;
        queues.values().any(|q| q.bytes >= self.cfg.flush.flush_bytes)
    }

    pub async fn flush_all(&self) -> Result<()> {
        let mut batches = Vec::<(String, Vec<Point>)>::new();
        {
            let mut queues = self.queues.lock().await;
            let keys: Vec<String> = queues.keys().cloned().collect();
            for k in keys {
                if let Some(q) = queues.get_mut(&k) {
                    if q.points.is_empty() {
                        continue;
                    }
                    let pts = std::mem::take(&mut q.points);
                    q.bytes = 0;
                    batches.push((k, pts));
                }
            }
        }

        for (key, points) in batches {
            self.flush_one(&key, &points).await?;
        }
        Ok(())
    }

    async fn flush_one(&self, key: &str, points: &[Point]) -> Result<()> {
        let mut parts = key.split('/');
        let day = parts.next().unwrap_or_default();
        let hour = parts.next().unwrap_or_default();
        let shard: u32 = parts.next().unwrap_or("0").parse().unwrap_or(0);
        let p = crate::storage::layout::PartitionInfo {
            start_ms: 0,
            day: day.to_string(),
            hour: hour.to_string(),
        };
        let seg_path = segment_path(&self.cfg.storage.data_dir, &p, shard);
        let idx_path = index_path(&self.cfg.storage.data_dir, &p, shard);
        if let Some(parent) = seg_path.parent() {
            tokio::fs::create_dir_all(parent).await?;
        }
        if let Some(parent) = idx_path.parent() {
            tokio::fs::create_dir_all(parent).await?;
        }

        let mut payload = Vec::new();
        let mut min_ts = i64::MAX;
        let mut max_ts = i64::MIN;
        let mut min_tag = u32::MAX;
        let mut max_tag = 0u32;
        for p in points {
            min_ts = min_ts.min(p.ts_epoch);
            max_ts = max_ts.max(p.ts_epoch);
            min_tag = min_tag.min(p.tag_id);
            max_tag = max_tag.max(p.tag_id);
            payload.extend_from_slice(&encode_segment_record(p));
        }
        let start_off = tokio::fs::metadata(&seg_path).await.map(|m| m.len()).unwrap_or(0);
        let mut seg = tokio::fs::OpenOptions::new()
            .create(true)
            .append(true)
            .open(&seg_path)
            .await?;
        seg.write_all(&payload).await?;
        let end_off = start_off + payload.len() as u64;

        if self.cfg.index.index_block_on_flush {
            let e = BlockIndexEntry {
                min_ts,
                max_ts,
                byte_offset_start: start_off,
                byte_offset_end: end_off,
                point_count: points.len() as u32,
                min_tag_id: min_tag,
                max_tag_id: max_tag,
            };
            let mut idx = tokio::fs::OpenOptions::new()
                .create(true)
                .append(true)
                .open(&idx_path)
                .await?;
            idx.write_all(&encode_block_index_entry(&e)).await?;
        }
        Ok(())
    }

    pub async fn stats(&self) -> WriterStats {
        self.stats.lock().await.clone()
    }
}

File: historian_v3/pkg/historian/writer.go
package historian

import (
	"os"
	"path/filepath"
	"sync"
	"time"
)

type WriterStats struct {
	AcceptedPoints    uint64 `json:"acceptedPoints"`
	DroppedPoints     uint64 `json:"droppedPoints"`
	DecodeErrors      uint64 `json:"decodeErrors"`
	SegmentSyncCount  uint64 `json:"segmentSyncCount"`
	SegmentSyncErrors uint64 `json:"segmentSyncErrors"`
}

type queueState struct {
	points []Point
	bytes  int
}

type HistorianWriter struct {
	cfg             Config
	lastStore       *LastValueStore
	wal             *WAL
	mu              sync.Mutex
	flushMu         sync.Mutex
	queues          map[string]*queueState
	stats           WriterStats
	stopCh          chan struct{}
	lastSegmentSync time.Time
}

func NewHistorianWriter(cfg Config, lastStore *LastValueStore, wal *WAL) *HistorianWriter {
	return &HistorianWriter{
		cfg:       cfg,
		lastStore: lastStore,
		wal:       wal,
		queues:    make(map[string]*queueState),
		stopCh:    make(chan struct{}),
	}
}

func (w *HistorianWriter) Start() {
	t := time.NewTicker(time.Duration(w.cfg.Flush.FlushIntervalMs) * time.Millisecond)
	go func() {
		for {
			select {
			case <-t.C:
				_ = w.FlushAll()
			case <-w.stopCh:
				t.Stop()
				return
			}
		}
	}()
}

func (w *HistorianWriter) Stop() {
	close(w.stopCh)
	_ = w.FlushAll()
}

func (w *HistorianWriter) RecoverFromWAL() error {
	if w.wal == nil {
		return nil
	}
	return w.wal.Replay(func(points []Point) error {
		return w.ingestBatchNoWAL(points)
	})
}

func (w *HistorianWriter) MarkDecodeError() {
	w.mu.Lock()
	w.stats.DecodeErrors++
	w.mu.Unlock()
}

func (w *HistorianWriter) Stats() WriterStats {
	w.mu.Lock()
	defer w.mu.Unlock()
	return w.stats
}

func (w *HistorianWriter) IngestBatch(points []Point) error {
	if w.wal != nil {
		if err := w.wal.Append(points); err != nil {
			return err
		}
	}
	return w.ingestBatchNoWAL(points)
}

func (w *HistorianWriter) ingestBatchNoWAL(points []Point) error {
	accepted := make([]Point, 0, len(points))
	needFlush := false
	w.mu.Lock()
	for _, p := range points {
		part := ComputePartition(p.TsEpoch, w.cfg)
		shard := ShardForTag(p.TagID, w.cfg.Storage.ShardCount)
		key := part.Day + "/" + part.Hour + "/" + itoa(shard)
		q := w.queues[key]
		if q == nil {
			q = &queueState{}
			w.queues[key] = q
		}
		if len(q.points) >= w.cfg.Flush.MaxQueuePoints {
			if w.cfg.Flush.BackpressurePolicy == "drop_new" {
				w.stats.DroppedPoints++
				continue
			}
			if len(q.points) > 0 {
				removed := SegmentRecordSize(q.points[0])
				q.points = q.points[1:]
				q.bytes -= removed
				if q.bytes < 0 {
					q.bytes = 0
				}
				w.stats.DroppedPoints++
			}
		}
		recLen := SegmentRecordSize(p)
		q.points = append(q.points, p)
		q.bytes += recLen
		w.stats.AcceptedPoints++
		accepted = append(accepted, p)
		if q.bytes >= w.cfg.Flush.FlushBytes {
			needFlush = true
		}
	}
	w.mu.Unlock()
	if len(accepted) > 0 {
		w.lastStore.UpdateBatch(accepted)
	}
	if needFlush {
		return w.FlushAll()
	}
	return nil
}

func (w *HistorianWriter) FlushAll() error {
	w.flushMu.Lock()
	defer w.flushMu.Unlock()
	return w.flushAllLocked()
}

func (w *HistorianWriter) flushAllLocked() error {
	type batch struct {
		key    string
		points []Point
		bytes  int
	}
	batches := make([]batch, 0, len(w.queues))
	w.mu.Lock()
	for k, q := range w.queues {
		if len(q.points) == 0 {
			continue
		}
		pts := q.points
		bytes := q.bytes
		q.points = nil
		q.bytes = 0
		batches = append(batches, batch{key: k, points: pts, bytes: bytes})
	}
	w.mu.Unlock()

	for _, b := range batches {
		if err := w.flushOne(b.key, b.points, b.bytes); err != nil {
			return err
		}
	}
	if w.wal != nil && len(batches) > 0 {
		rotated, err := w.wal.Rotate()
		if err == nil && rotated != "" {
			_ = os.Remove(rotated)
		}
	}
	return nil
}

func (w *HistorianWriter) RunMaintenance(fn func() error) error {
	w.flushMu.Lock()
	defer w.flushMu.Unlock()
	if err := w.flushAllLocked(); err != nil {
		return err
	}
	return fn()
}

func (w *HistorianWriter) flushOne(key string, points []Point, payloadBytes int) error {
	day, hour, shard := parseQueueKey(key)
	p := PartitionInfo{Day: day, Hour: hour}
	segPath := SegmentPath(w.cfg.Storage.DataDir, p, shard)
	idxPath := IndexPath(w.cfg.Storage.DataDir, p, shard)
	if err := os.MkdirAll(filepath.Dir(segPath), 0o755); err != nil {
		return err
	}
	if err := os.MkdirAll(filepath.Dir(idxPath), 0o755); err != nil {
		return err
	}
	if payloadBytes <= 0 {
		for i := range points {
			payloadBytes += SegmentRecordSize(points[i])
		}
	}
	payload := make([]byte, payloadBytes)
	payloadOff := 0
	minTs := int64(^uint64(0) >> 1)
	maxTs := int64(-1 << 63)
	minTag := uint32(^uint32(0))
	maxTag := uint32(0)
	for _, point := range points {
		if point.TsEpoch < minTs {
			minTs = point.TsEpoch
		}
		if point.TsEpoch > maxTs {
			maxTs = point.TsEpoch
		}
		if point.TagID < minTag {
			minTag = point.TagID
		}
		if point.TagID > maxTag {
			maxTag = point.TagID
		}
		payloadOff += encodeSegmentRecordTo(payload[payloadOff:], point)
	}
	if payloadOff != len(payload) {
		payload = payload[:payloadOff]
	}
	startOff := uint64(0)
	if st, err := os.Stat(segPath); err == nil {
		startOff = uint64(st.Size())
	}
	sf, err := os.OpenFile(segPath, os.O_CREATE|os.O_APPEND|os.O_WRONLY, 0o644)
	if err != nil {
		return err
	}
	if _, err := sf.Write(payload); err != nil {
		_ = sf.Close()
		return err
	}
	w.syncSegmentFileByPolicy(sf)
	_ = sf.Close()
	endOff := startOff + uint64(len(payload))

	if w.cfg.Index.IndexBlockOnFlush {
		entry := BlockIndexEntry{
			MinTs: minTs, MaxTs: maxTs,
			ByteOffsetStart: startOff, ByteOffsetEnd: endOff,
			PointCount: uint32(len(points)),
			MinTagID:   minTag, MaxTagID: maxTag,
		}
		iff, err := os.OpenFile(idxPath, os.O_CREATE|os.O_APPEND|os.O_WRONLY, 0o644)
		if err != nil {
			return err
		}
		if _, err := iff.Write(EncodeBlockIndexEntry(entry)); err != nil {
			_ = iff.Close()
			return err
		}
		w.syncSegmentFileByPolicy(iff)
		_ = iff.Close()
	}
	return nil
}

func parseQueueKey(k string) (day string, hour string, shard int) {
	parts := split3(k)
	return parts[0], parts[1], atoi(parts[2])
}

func (w *HistorianWriter) syncSegmentFileByPolicy(f *os.File) {
	needSync := false
	switch w.cfg.FSync.SegmentPolicy {
	case "always":
		needSync = true
	case "interval":
		needSync = time.Since(w.lastSegmentSync) >= time.Duration(w.cfg.FSync.SegmentIntervalMs)*time.Millisecond
	case "off":
		needSync = false
	default:
		needSync = time.Since(w.lastSegmentSync) >= time.Duration(w.cfg.FSync.SegmentIntervalMs)*time.Millisecond
	}
	if !needSync {
		return
	}
	if err := f.Sync(); err != nil {
		w.mu.Lock()
		w.stats.SegmentSyncErrors++
		w.mu.Unlock()
		return
	}
	w.lastSegmentSync = time.Now()
	w.mu.Lock()
	w.stats.SegmentSyncCount++
	w.mu.Unlock()
}
